backend/persistent/plotly-api.js

```diff
@@ -0,0 +1,254 @@
+        method: 'GET',
+        username,
+        apiKey,
+        accessToken
+    });
+}
+
+export function deleteGrid(fid, requestor) {
+    const {username, apiKey, accessToken} = getCredentials(requestor);
+
+    return plotlyAPIRequest(`grids/${fid}`, {
+        method: 'DELETE',
+        username,
+        apiKey,
+        accessToken
+    });
+}
+
+export function updateGrid(rows, fid, uids, requestor) {
+    const {username, apiKey, accessToken} = getCredentials(requestor);
+
+    // TODO - Test case where no rows are returned.
+    if (uids.length !== (rows[0] || []).length) {
+        Logger.log(`
+            A different number of columns was returned in the
+            query than what was initially saved in the grid.
+            ${rows[0].length} columns were queried,
+            ${uids.length} columns were originally saved.
+            The connector does not create columns (yet),
+            and so we will only update the first ${uids.length}
+            columns.
+        `);
+    }
+
+    /*
+     * For now, only update up to the number of columns that are
+     * already saved. In the future, we should just create more
+     * columns. See error message above.
+     */
+    const columns = getColumns(rows, uids.length);
+
+    const url = `grids/${fid}/col?uid=${uids.join(',')}`;
+    const body = {
+        cols: JSON.stringify(columns.map(column => ({
+            data: column
+        })))
+    };
+    return plotlyAPIRequest(url, {body, username, apiKey, accessToken, method: 'PUT'});
+}
+
+function getColumns(rows, maxColumns) {
+    const columns = [];
+
+    // don't return more than maxColumns
+    const columnsLength = Math.min(maxColumns, (rows[0] || []).length);
+
+    for (let i = 0; i < columnsLength; i++) {
+        columns.push([]);
+    }
+
+    for (let rowIndex = 0; rowIndex < rows.length; rowIndex++) {
+        for (let columnIndex = 0; columnIndex < columnsLength; columnIndex++) {
+            columns[columnIndex][rowIndex] = rows[rowIndex][columnIndex];
+        }
+    }
+
+    return columns;
+}
+
+// Resolve if the requestor has permission to update fid, reject otherwise
+export function checkWritePermissions(fid, requestor) {
+    const {username, apiKey, accessToken} = getCredentials(requestor);
+
+    // Check if the user even exists
+    if (!username || !(apiKey || accessToken)) {
+        /*
+         * Warning: The front end looks for "Unauthenticated" in this error message. Don't change it!
+         */
+        const errorMessage = (
+            `Unauthenticated: Attempting to update grid ${fid} but the ` +
+            `authentication credentials for the user "${requestor}" do not exist.`
+        );
+        Logger.log(errorMessage, 0);
+        throw new Error(errorMessage);
+    }
+
+    return plotlyAPIRequest(`grids/${fid}`, {
+        username: requestor,
+        apiKey,
+        accessToken,
+        method: 'GET'
+    }).then(function(res) {
+        if (res.status === 404) {
```

Comment:
We should maybe flip the codes into constants with names?

Reply:
No, it'd make the code less readable (in this context it's clear 404 is an HTTP status code).

Reply:
OK

Reply:
@gzork One can always open a node session and run:
```
> http.STATUS_CODES[409]
'Conflict'
```

I can't think of any REST API that isn't defined in terms of numeric status codes. Let's say, I'm doing a review, and I come across an uncommon status code written like: `if (res.status === CONFLICT) {...}`. Then, I need to stop my review, and find where `CONFLICT` has been defined, to be able to confirm that the right status code has been used.